lib/server.js

```diff
@@ -1223,7 +1070,122 @@ Server.prototype._setupRequest = function _setupRequest(req, res) {
     if (self.name !== '') {
         res.setHeader('Server', self.name);
     }
-    res.version = self.router.versions[self.router.versions.length - 1];
+
+    // Request lifecycle events
+    // attach a listener for 'close' events, this will let us set
+    // a flag so that we can stop processing the request if the client closes
+    // the connection (or we lose the connection).
+    function onClose() {
+        res.removeListener('finish', onFinish);
+        res.removeListener('error', onError);
+        req._connectionState = 'close';
+
+        self._finishReqResCycle(req, res, new errors.RequestCloseError());
+    }
+    req.once('close', onClose);
+
+    // Response lifecycle events
+    function onFinish() {
+        req.removeListener('close', onClose);
+        res.removeListener('error', onError);
+        self._finishReqResCycle(req, res, res.err);
+    }
+    function onError(err) {
+        req.removeListener('close', onClose);
+        res.removeListener('finish', onFinish);
+        self._finishReqResCycle(req, res, err);
+    }
+    res.once('finish', onFinish);
```

Comment:
Is this a breaking change from 6.x? IIRC in 6.x we consider a request finished once it has reached the end of the handler chain. Using this pattern, it won't fire until the request is completely done. 

I have to think about whether this is better/worse/different, but definitely has an impact on how we might view metrics. cc @yunong